### PR TITLE
fix(s3): omit checksum headers from GetObject/HeadObject unless x-amz-checksum-mode: ENABLED

### LINKED ISSUE #1632: [BUG] s3: PHP fopen on S3 objects fails with "Stream is not seekable"
## Summary

`fopen('s3://bucket/key', 'r')` in PHP fails with `Stream is not seekable` when using Floci as the S3 endpoint.

> **Disclosure**: the bug was discovered manually. The root cause analysis in this report was produced with AI assistance.

## Affected versions

Confirmed on **1.5.26**. The underlying cause has been present since early versions and is not fixed in 1.5.27 or 1.5.28.

### 1. Install the AWS SDK for PHP

The bug requires **aws/aws-sdk-php >= 3.130** (when checksum validation was introduced). Older versions silently ignore the unexpected checksum headers and do not trigger the failure.

```bash
composer require "aws/aws-sdk-php:^3.300"
```

### 2. Create the test script

```php
<?php
require __DIR__ . '/vendor/autoload.php';

use Aws\S3\S3Client;
use Aws\S3\StreamWrapper;

$s3 = new S3Client([
    'version'                 => 'latest',
    'region'                  => 'us-east-1',
    'endpoint'                => 'http://localhost:4566',
    'credentials'             => ['key' => 'test', 'secret' => 'test'],
    'use_path_style_endpoint' => true,
]);

StreamWrapper::register($s3);

// Setup: create bucket and upload a test object
$s3->createBucket(['Bucket' => 'test-bucket']);
$s3->putObject([
    'Bucket'      => 'test-bucket',
    'Key'         => 'test.txt',
    'Body'        => 'Hello, world!',
    'ContentType' => 'text/plain',
]);

// This fails with: Stream is not seekable
$stream = fopen('s3://test-bucket/test.txt', 'r');
var_dump($stream); // bool(false)
```

### 3. Run against Floci

```bash
php test.php
```

Expected output:
```
resource(5) of type (stream)
```

Actual output:
```
Warning: Stream is not seekable in .../aws-sdk-php/src/S3/StreamWrapper.php
Warning: fopen(s3://test-bucket/test.txt): Failed to open stream: "Aws\S3\StreamWrapper::stream_open" call failed in test.php on line 27
bool(false)
```

## Root cause

What follows it an analysis done using Claude Code. I am not a PHP developer. I was tasked with migrating from localstack to floci and there are some PHP applications in our stack. I know just enough to create a test php script that can reproduce the issue but I cannot do (at least efficiently) an in-depth analysis. What follows might be total BS, i don't know. On a surface level it makes sense but I cannot be completly sure.

Floci unconditionally includes `x-amz-checksum-crc32` (and other `x-amz-checksum-*` headers) in **every** `GetObject 200` response, regardless of whether the client requested checksum validation.

**AWS real behavior**: checksum headers are only returned in `GetObject` responses when the client sends `x-amz-checksum-mode: ENABLED`. Without that request header, AWS does not include any `x-amz-checksum-*` header in the response.

**What the PHP SDK does when it sees a checksum header**: `ValidateResponseChecksumResultMutator` intercepts the response, reads the entire body to compute the checksum, and then calls `seek(0)` to rewind the stream for the caller. The PHP S3 `StreamWrapper` opens the connection in streaming mode (`$command['@http']['stream'] = true`), which returns the HTTP body as a live socket stream — not seekable. The `seek(0)` call therefore throws `RuntimeException: Stream is not seekable`, which `boolCall` catches and surfaces as the warning above.

Stack trace (abbreviated):

```
RuntimeException: Stream is not seekable
  GuzzleHttp\Psr7\Stream::seek()
  GuzzleHttp\Psr7\Utils::hash(..., 'crc32b', rewind=true)
  ValidateResponseChecksumResultMutator::getEncodedValue()
  ValidateResponseChecksumResultMutator::validateChecksum()
  S3Parser::executeS3ResultMutators()
  [inside StreamWrapper::openReadStream() → boolCall → triggerError]
```

## Location of the fix

`S3Controller.java`, method `getObject`. It currently calls:

```java
return fullObjectResponse(bucket, key, versionId, obj, overrides);
```

which resolves to `appendObjectHeaders(resp, obj, overrides, /* includeChecksum= */ true)`.

The fix is to read the `x-amz-checksum-mode` request header and only pass `includeChecksum=true` when the client explicitly requested it:

```java
@HeaderParam("x-amz-checksum-mode") String checksumMode,
// ...
boolean includeChecksum = "ENABLED".equalsIgnoreCase(checksumMode);
return fullObjectResponse(bucket, key, versionId, obj, overrides, includeChecksum);
```

The same gate was already applied correctly for range (206) responses in #1040.

### PR DESCRIPTION
**Pull request by mihaiblidaru**
## Problem

Closes #1632

`GetObject` and `HeadObject` responses unconditionally include `x-amz-checksum-*` headers regardless of whether the request contained `x-amz-checksum-mode: ENABLED`. On real AWS, these headers are only returned when the client explicitly opts in via that header.

This causes `fopen('s3://bucket/key', 'r')` in PHP's AWS SDK to fail with **"Stream is not seekable"**: the SDK's `ValidateResponseChecksumResultMutator` detects a checksum header in the response, reads the entire body to validate it, then calls `seek(0)` to rewind — which fails because the underlying stream is a non-seekable socket (the SDK uses `stream=true` internally to avoid buffering).

## Root cause

`appendObjectHeaders` / `appendChecksumHeaders` were called unconditionally in both `getObject` and `headObject`, without checking for the `x-amz-checksum-mode` request header.

## Fix

- Added `@HeaderParam("x-amz-checksum-mode")` to `getObject` and `headObject`
- Gate checksum header inclusion behind `"ENABLED".equalsIgnoreCase(checksumMode)`
- Added `includeChecksum` boolean parameter to `fullObjectResponse` to propagate the gate through range-request edge cases
- Updated integration tests: existing `getObject`/`headObject` tests now assert checksum headers are absent; new tests assert they are present when `x-amz-checksum-mode: ENABLED` is sent

## Testing

All 117 integration tests pass. Verified end-to-end with PHP 8.2 + `aws/aws-sdk-php`:
- Against unpatched 1.5.26: `fopen` fails with "Stream is not seekable"
- Against this fix: `fopen`, `fread`, and `fseek` all succeed

### CODE REVIEW

File: src/test/java/io/github/hectorvent/floci/services/s3/S3IntegrationTest.java [hunk 1]
```diff
@@ -87,10 +87,22 @@ void getObject() {
...
             .header("x-amz-meta-owner", equalTo("team-a"))
             .header("x-amz-storage-class", equalTo("STANDARD_IA"))
-            .header("x-amz-checksum-crc64nvme", notNullValue())
+            .header("x-amz-checksum-crc64nvme", nullValue())
             .body(equalTo("Hello World from S3!"));
     }
 
+    @Test
+    @Order(5)
+    void getObjectWithChecksumModeReturnsChecksum() {
```
[the remark this answers is not in the document]
Comment (mihaiblidaru) [reply src/test/java/io/github/hectorvent/floci/services/s3/S3IntegrationTest.java]: Fixed in 12d18e2c. Rather than renumber the whole block, I moved both new tests to the **end** of the 5–22 `@Order` block with unique `@Order(23)`/`@Order(24)` (the block max was 22, next block starts at 30, so these are free). `getObject` and `headObject` keep their original `@Order(5)`/`@Order(7)`, now unambiguous — and no existing test was renumbered.

**Comment by mihaiblidaru** [comment]
Thanks for the review — both P2 items are addressed in 12d18e2c:

**Dead-code footgun (`appendObjectHeaders` 3-arg overload).** Removed it. Since all call sites now pass `includeChecksum` explicitly to the 4-arg version, the 3-arg overload (and the transitively-dead 2-arg overload, plus the now-orphaned `ResponseHeaderOverrides.NONE`) were unreachable. Deleting them leaves only the explicit 4-arg version, so there's no default-`true` path that could silently re-introduce the bug for a future operation.

**Duplicate `@Order` values.** Resolved — see the inline thread. Moved both new tests to the end of the block with unique `@Order(23)`/`@Order(24)`.

All 117 `S3IntegrationTest` cases still pass.

**Comment by mihaiblidaru** [comment]
The `@Order` parameter values in `S3IntegrationTest.java` doesn't seem to follow any logic. I moved the two tests I added to the first gap I found. There are many other duplicated @Order values. If this it's not an issue maybe the Greptile bot shoud be instructed to ignore this kind of issues

**Comment by hectorvent** [comment]
This one is in great shape, thanks. Verified against [service-2.json](https://github.com/boto/botocore/blob/develop/botocore/data/s3/2006-03-01/service-2.json) (s3/2006-03-01): `ChecksumMode` is exactly the `x-amz-checksum-mode` header on both operations and the model doc says "To retrieve the checksum, this mode must be enabled", so the gate matches the documented contract. PutObject and GetObjectAttributes were correctly left alone, the PHP SDK root cause analysis checks out, and both header directions are now locked by tests.

One optional nit: in `handleRangeRequest` the empty object suffix range path calls `fullObjectResponse(..., false)` even when the client sent checksum mode. Since that path returns a full 200 rather than a 206, real S3 would include the checksum there. It is an empty object corner case, so feel free to ignore or thread the flag through:

```java
return fullObjectResponse(bucket, key, versionId, obj, overrides, includeChecksum);
```

And one housekeeping request: could you please update the commit history to remove the `Co-authored-by: Claude <...>` line from the commit message? We prefer to keep attribution focused on human contributors for this project (it is also in the contribution guidelines), and the squash merge would otherwise carry it into main.

**Comment by mihaiblidaru** [comment]
Addressed both points:
  
  - **Fixed the edge case**: `includeChecksum` is now computed once and threaded through `handleRangeRequest`, so the empty-object suffix-range path honors `x-amz-checksum-mode: ENABLED` instead of hardcoding `false`. Added a dedicated test for this edge case plus an assertion on the existing suffix-range test confirming the checksum is omitted when the header isn't sent. 
  - **Rewrote the commit history** to drop the `Co-authored-by: Claude` trailer and rebased cleanly onto latest `main`.

**Comment by mihaiblidaru** [comment]
Hi @hectorvent,

I've rebased this branch onto `main` to resolve the merge conflicts. Wanted to flag a couple of things that came up during the rebase in case they're useful context for review:

**Test ordering / dependencies in `S3IntegrationTest`**

This test class relies on `@TestMethodOrder(OrderAnnotation.class)` with fairly tight coupling between tests — several tests share a bucket/object created by an earlier-ordered test and rely on `cleanupAndDeleteBucket` (a later-ordered test) to tear it down. That made a couple of things tricky:

- New tests added upstream since this PR was opened (`copyObject`, `copyObjectWithoutChecksumOverride`, `copyObjectPreservesNonDefaultChecksum`) assert that `GetObject` returns checksum headers unconditionally. That's no longer true after this PR's fix (checksums are only returned when `x-amz-checksum-mode: ENABLED` is sent), so I added that header to their verification requests.
- A failed assertion mid-test skips that test's own cleanup step, which can leave objects behind in the shared bucket — this showed up as an unrelated `409` on `cleanupAndDeleteBucket` when the real failure was the missing checksum-mode header a few tests earlier.
- For the two new tests this PR adds (`getObjectWithChecksumModeReturnsChecksum`, `headObjectWithChecksumModeReturnsChecksum`), I made them fully self-contained (own bucket, own cleanup) rather than relying on the shared bucket, and gave them `@Order` values that don't collide with any existing ones.

Let me know if anything else is needed on my end.

**Comment by mihaiblidaru** [comment]
The new "Docs Action Tables" check fails because my branch is not up to date with the `main` branch. Do I rebase again?

**Comment by mihaiblidaru** [comment]
Rebased again. Runned `docs-test`, `docs-check` and `docs-sync` and everything seems up to date.

```bash
[floci]$ make docs-test
python3 -m pytest tools/docs -q
....................                                                                                                                                                                                                                                                         [100%]
20 passed in 0.03s
[floci]$ make docs-check
[floci]$ make docs-sync
python3 tools/docs/regen_action_docs.py
[floci]$ git diff
[floci]$ git status
On branch fix/s3-checksum-mode-gate
Your branch is up to date with 'origin/fix/s3-checksum-mode-gate'.

nothing to commit, working tree clean
[floci]$ 

```

**Review by hectorvent — APPROVED** [review]
Thanks @mihaiblidaru
